### ISSUE TITLE
[google_maps_flutter] Explain how GoogleMapController works

### ISSUE THREAD

**Issue by flutter2k20** [flutter/flutter#74345]
Hi,

I am working on a google maps based Flutter application. I am using google_maps_flutter plugin for displaying google maps. 

Looking at the examples, a `GoogleMapController` is passed to the `onMapCreated` function, and a reference is kept with in the class. e.g. 
https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter/example/lib/map_click.dart
But I don't see anywhere in the class that the controller is disposed off or released. 

Also, at the following example, `GoogleMapsController` is accessed completely differently by using `Completer` `Completer<GoogleMapController> _controller = Completer();`
https://github.com/flutter/plugins/tree/master/packages/google_maps_flutter/google_maps_flutter 

Can someone from the plugin team look at the discrepancy and let us know 

a. What is the CORRECT way to access `GoogleMapController`?
b. If `GoogleMapController` should be disposed off or not?

Thanks!

**Comment by stuartmorgan** [comment]
> But I don't see anywhere in the class that the controller is disposed off or released.

That example page is using the map for as long as it's displayed, so needs the controller.

> Also, at the following example, `GoogleMapsController` is accessed completely differently by using `Completer` `Completer<GoogleMapController> _controller = Completer();`

It's accessed the same way, via `onMapCreated`. The completer is used for convenience of blocking functionality on the async return. This isn't specific to maps, it's just one possible pattern for dealing with async access to an object.

> Can someone from the plugin team look at the discrepancy and let us know
> a. What is the CORRECT way to access `GoogleMapController`?

There is no discrepancy here; you get the controller in the `onMapCreated` callback. How you manage that async access in your application is up to you, and may depend on your use case.

> b. If `GoogleMapController` should be disposed off or not?

It's not clear to me what your question is. As with any object, you should keep it for as long as you need it.

**Comment by wazasoft** [comment]
> > But I don't see anywhere in the class that the controller is disposed off or released.
> 
> That example page is using the map for as long as it's displayed, so needs the controller.
> 
> > Also, at the following example, `GoogleMapsController` is accessed completely differently by using `Completer` `Completer<GoogleMapController> _controller = Completer();`
> 
> It's accessed the same way, via `onMapCreated`. The completer is used for convenience of blocking functionality on the async return. This isn't specific to maps, it's just one possible pattern for dealing with async access to an object.
> 
> > Can someone from the plugin team look at the discrepancy and let us know
> > a. What is the CORRECT way to access `GoogleMapController`?
> 
> There is no discrepancy here; you get the controller in the `onMapCreated` callback. How you manage that async access in your application is up to you, and may depend on your use case.
> 
> > b. If `GoogleMapController` should be disposed off or not?
> 
> It's not clear to me what your question is. As with any object, you should keep it for as long as you need it.

Hi Stuart,

Thanks for the reply, but some how you completely missed the point. 

Let me try again.

There is a public function `dispose` in the `GoogleMapsController`.

```
 /// Disposes of the platform resources
  void dispose() {
    GoogleMapsFlutterPlatform.instance.dispose(mapId: mapId);
  }
```

 Do we (as in whoever uses this plugin and gets the controller passed as a parameter in `onMapCreated` callback) call that `dispose` function ourself when we no longer need the map or the controller's dispose gets called "automagically" somehow?

**Comment by flutter2k20** [comment]
Hi Stuart,

Thanks for the reply, but some how you completely missed the point. 

Let me try again.

There is a public function `dispose` in the `GoogleMapsController`.

```
 /// Disposes of the platform resources
  void dispose() {
    GoogleMapsFlutterPlatform.instance.dispose(mapId: mapId);
  }
```

 Do we (as in whoever uses this plugin and gets the controller passed as a parameter in `onMapCreated` callback) call that `dispose` function ourself when we no longer need the map or the controller's dispose gets called "automagically" somehow?

**Comment by stuartmorgan** [comment]
> Thanks for the reply, but some how you completely missed the point.

You were using the words "disposed" and "released" interchangeably, and taking about " keeping a reference", which is why it was not clear to me you were talking about the `dispose` method rather than object lifecycle.

> Do we (as in whoever uses this plugin and gets the controller passed as a parameter in `onMapCreated` callback) call that `dispose` function ourself when we no longer need the map or the controller's dispose gets called "automagically" somehow?

It's called by the map widget; you do not need to call it yourself in normal usage. That's why the examples don't call it.
